packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +51,400 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
-      primaryKey: arg.primaryKey,
-      asymmetric: arg.asymmetric || false,
-      embedded: arg.embedded || false,
-      properties: Object.fromEntries(
-        Object.entries(arg.properties).map(([name, property]) => {
-          const canonicalPropertySchema = normalizePropertySchema(name, property);
-          // A primary key is always indexed
-          if (name === arg.primaryKey) {
-            canonicalPropertySchema.indexed = true;
-          }
-          validateCanonicalPropertySchema(arg.name, canonicalPropertySchema);
-          return [name, canonicalPropertySchema];
-        }),
-      ),
-    };
+    throw new Error("Array of properties are no longer supported. Use an object instead.");
   }
+  // --------------------------------------------------
+
+  return {
+    name: arg.name,
+    primaryKey: arg.primaryKey,
+    asymmetric: !!arg.asymmetric,
+    embedded: !!arg.embedded,
+    properties: normalizePropertySchemas(arg.name, arg.properties, arg.primaryKey),
+  };
+}
+
+/**
+ * Transform user-provided property schemas into their canonical forms.
+ */
+function normalizePropertySchemas(
+  objectName: string,
+  schemas: PropertiesTypes,
+  primaryKey?: string,
+): Record<string, CanonicalObjectSchemaProperty> {
+  const normalizedSchemas: Record<string, CanonicalObjectSchemaProperty> = {};
+  for (const propertyName in schemas) {
+    normalizedSchemas[propertyName] = normalizePropertySchema(
+      { objectName, propertyName },
+      schemas[propertyName],
+      primaryKey === propertyName,
+    );
+  }
+
+  return normalizedSchemas;
 }
 
-// TODO: Extend to handle all short hands
+/**
+ * Transform a user-provided property schema into its canonical form.
+ */
 export function normalizePropertySchema(
-  name: string,
+  name: ObjectAndPropertyName,
   schema: string | ObjectSchemaProperty,
+  isPrimaryKey = false,
 ): CanonicalObjectSchemaProperty {
-  if (typeof schema === "string") {
-    return normalizePropertySchema(name, normalizePropertyType(schema));
-  } else {
-    const { type, ...rest } = schema;
-    const propertySchemaFromType = normalizePropertyType(type, typeof rest.objectType !== "string");
-    // Type casting, since it is expected that `normalizePropertyType` moves an object linked type into `objectType`
-    const mergedSchema = {
-      ...removeUndefinedValues(propertySchemaFromType),
-      ...removeUndefinedValues(rest as Omit<ObjectSchemaProperty, "name">),
-    } as ObjectSchemaProperty & { type: PropertyTypeName };
-    if (mergedSchema.type === "mixed" || mergedSchema.objectType === "mixed") {
-      assert(mergedSchema.optional !== false, "Mixed values should be declared as optional");
+  const normalizedSchema =
+    typeof schema === "string"
+      ? normalizePropertySchemaString(name, schema)
+      : normalizePropertySchemaObject(name, schema);
+
+  if (isPrimaryKey) {
+    assert(!normalizedSchema.optional, errMessage(name, "Optional properties cannot be used as a primary key."));
```

Comment:
Realm Kotlin (and Realm Core) allows `null` as value for the primary key. The question is if we want to allow that too?

Reply:
What would be the use case for `null`?